x-pack/plugins/actions/common/validate_email_addresses.ts

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { parseAddressList } from 'email-addresses';
+import { ValidatedEmail, InvalidEmailReason } from './types';
+import { hasMustacheTemplate } from './mustache_template';
+
+export interface ValidateEmailAddressesOptions {
+  allowMustache?: boolean;
+}
+
+// this can be useful for cases where a plugin needs this function,
+// but the actions plugin may not be available.  This could be used
+// as a stub for the real implementation.
+export function validateEmailAddressesAsAlwaysValid(
+  addresses: string[],
+  options?: ValidateEmailAddressesOptions
+): ValidatedEmail[] {
+  return addresses.map((address) => ({ address, valid: true }));
+}
+
+export function validateEmailAddresses(
+  allowedDomains: string[] | null,
+  addresses: string[],
+  options: ValidateEmailAddressesOptions = {}
+): ValidatedEmail[] {
+  // note: this is the legacy default, which would in theory allow
+  // mustache strings, so options.allowMustache is ignored in this
+  // case - everything is valid!
+  if (allowedDomains == null) {
+    return validateEmailAddressesAsAlwaysValid(addresses);
+  }
+
+  return addresses.map((address) => validateEmailAddress(allowedDomains, address, options));
+}
+
+export function invalidEmailsAsMessage(validatedEmails: ValidatedEmail[]): string | undefined {
+  const invalid = validatedEmails.filter(
+    (validated) => !validated.valid && validated.reason === InvalidEmailReason.invalid
+  );
+  const notAllowed = validatedEmails.filter(
+    (validated) => !validated.valid && validated.reason === InvalidEmailReason.notAllowed
+  );
+
+  const messages: string[] = [];
+  if (invalid.length !== 0) {
+    messages.push(`not valid emails: ${addressesFromValidatedEmails(invalid).join(', ')}`);
+  }
+  if (notAllowed.length !== 0) {
+    messages.push(`not allowed emails: ${addressesFromValidatedEmails(notAllowed).join(', ')}`);
+  }
+
+  if (messages.length === 0) return;
+
+  return messages.join('; ');
+}
+
+// in case the npm email-addresses returns unexpected things ...
+function validateEmailAddress(
+  allowedDomains: string[],
+  address: string,
+  options: ValidateEmailAddressesOptions
+): ValidatedEmail {
+  if (options.allowMustache && hasMustacheTemplate(address)) {
+    return { address, valid: true };
+  }
```

Comment:
We spoke offline about this:

Anything that allows you to short-circuit a validation function is scary 😅 
In this case it's justified: the same function is used when a connector is first _added_, and again later when the connector is _executed_.
The former will allow mustache templates to essentially pass validation (because there is no data), but the latter will not, because at that point the field has been rendered to a string with the actual data.

That said, I think it would be good to add a comment here explaining the reasoning behind this for future maintainers.

Reply:
good idea, thx!